File: examples/review_diff_package.py
import os
from pathlib import Path

from forma import OpenAIResponsesProvider, agent
from review_diff_forma import ReviewDiffOutput, assert_review_diff_output


review_diff = agent(
    file=Path("examples/review_diff.forma"),
    task="review_diff",
    provider=OpenAIResponsesProvider(
        api_key=os.environ["OPENAI_API_KEY"],
        model=os.environ.get("OPENAI_MODEL", "gpt-5"),
    ),
)


def review_code_diff(diff: str) -> ReviewDiffOutput:
    result = review_diff.run({"diff": diff, "max_findings": 5})
    if not result.ok:
        raise RuntimeError(result.error or "Forma review_diff failed")
    return assert_review_diff_output(result.output)

File: examples/review_diff_package.ts
import { readFileSync } from "node:fs";
import { OpenAIResponsesProvider, agent } from "@forma-lang/forma";
import { assertReviewDiffOutput, type ReviewDiffOutput } from "./review_diff.forma.js";

interface ProviderProfile {
  provider: string;
  model: string;
  apiKeyEnv: string;
}

function loadProviderProfile(): ProviderProfile {
  return JSON.parse(readFileSync("examples/forma.provider.json", "utf8")) as ProviderProfile;
}

const providerProfile = loadProviderProfile();
if (providerProfile.provider !== "openai-responses") {
  throw new Error(`Unsupported Forma provider: ${providerProfile.provider}`);
}

const reviewDiff = agent({
  file: "examples/review_diff.forma",
  task: "review_diff",
  provider: new OpenAIResponsesProvider({
    apiKey: process.env[providerProfile.apiKeyEnv] ?? "",
    model: providerProfile.model,
  }),
});

export async function reviewCodeDiff(diff: string): Promise<ReviewDiffOutput> {
  const result = await reviewDiff.run({ diff, max_findings: 5 });
  if (!result.ok) {
    throw new Error(result.error ?? "Forma review_diff failed");
  }
  return assertReviewDiffOutput(result.output);
}
